### PR TITLE
[OdomSubscriber] make odom_topic param dynamic

### PR DESCRIPTION
**Pull request by doisyg**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | https://github.com/ros-planning/navigation2/issues/956 |
| Primary OS tested on | Ubuntu|
| Robotic platform tested on | Wyca Elodie simulation |

---

## Description of contribution in a few bullet points
Make the odom_topic parameter of OdomSubscriber dynamic. I have a personal need of changing at runtime the odometry source used by nav2, so sharing here.
TBD: relevance and test.

## Description of documentation updates required from your changes
Mention in migration guide

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Linter

> Add #include <vector> for vector<>

File: nav2_dwb_controller/nav_2d_utils/include/nav_2d_utils/odom_subscriber.hpp [hunk 1]
```diff
@@ -91,9 +97,38 @@ class OdomSubscriber
...
+    return result;
+  }
+
   rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
   nav_2d_msgs::msg::Twist2DStamped odom_vel_;
   std::mutex odom_mutex_;
+  nav2_util::LifecycleNode::WeakPtr nh_;
+  std::string odom_topic_;
+  // Dynamic parameters handler
+  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr dyn_params_handler;
```
Comment (SteveMacenski) [inline nav2_dwb_controller/nav_2d_utils/include/nav_2d_utils/odom_subscriber.hpp:131]: trailing underscore for class member

**Comment by SteveMacenski** [comment]
This seems really quite odd. It's not odd enough that I wouldn't merge it, so I would question the design of a system that requires it and perhaps we can find you a better solution. 

Why not use the typical topic muxing methodology? Have a `output_odom` topic that you have something that toggles your preferred odometry source to publish to and have the system subscribe to that? Runtime changing subscribers breaks a bit of the lifecycle readiness guarantees you get when we create the network interfaces before the active state. Further, you still need to have some autonomy code that is changing this dynamic parameter, that could be trivially changed to just pushing to the `output` topic instead and no need to change any subscriptions at runtime. 

Also, needs test coverage on dynamic parameters.

**Comment by doisyg** [comment]
Hum, a topic muxer would indeed make more sense. Let me reflect on that, but I believe you're right and this is not needed.

**Comment by doisyg** [comment]
For reference, the generic topic muxer node has not been ported to ROS2 yet, see: https://github.com/ros-tooling/topic_tools/issues/3

**Comment by doisyg** [comment]
Closing, after testing, a custom odom muxer makes much more sense.

**Pull request closed without merging**